### PR TITLE
Allow default values for `ref readonly` parameters

### PR DESCRIPTION
**Pull request by jjonescz**
Speclet: https://github.com/dotnet/csharplang/blob/05ed9dcfba84a038e02301a9b267516edfe14159/proposals/ref-readonly-parameters.md#default-parameter-values
Test plan: https://github.com/dotnet/roslyn/issues/68056

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 1]
```diff
@@ -2873,6 +2873,322 @@ static void Main()
...
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (4,20): error CS1745: Cannot specify default parameter value in conjunction with DefaultParameterAttribute or OptionalAttribute
+            //     static void M([DefaultParameterValue(1)] ref readonly int i = 1) => throw null;
+            Diagnostic(ErrorCode.ERR_DefaultValueUsedWithAttributes, "DefaultParameterValue").WithLocation(4, 20),
+            // (4,67): warning CS9521: A default value is specified for 'ref readonly' parameter 'i', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'.
+            //     static void M([DefaultParameterValue(1)] ref readonly int i = 1) => throw null;
+            Diagnostic(ErrorCode.WRN_RefReadonlyParameterDefaultValue, "1").WithArguments("i").WithLocation(4, 67),
+            // (4,67): error CS8017: The parameter has multiple distinct default values.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2972]: >// (4,67): error CS8017: The parameter has multiple distinct default values. [](http://example.com/codeflow?start=12&length=76)

This is probably unrelated to the changes, but why is this error reported?

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 2]
```diff
@@ -2873,6 +2873,322 @@ static void Main()
...
+                {
+                    M();
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (8,9): error CS7036: There is no argument given that corresponds to the required parameter 'd' of 'C.M(ref readonly DateTime)'
+            //         M();
+            Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "M").WithArguments("d", "C.M(ref readonly System.DateTime)").WithLocation(8, 9));
+    }
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:3190]: > } [](http://example.com/codeflow?start=3&length=2)

It looks like we are not testing consumption of default values from metadata. I think we should cover at least all valid forms of encoding: attributes only, constant value, combination of both.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2988]: > I guess the latter should not report if any of the values is bad

Probably. Could you open an issue for this?

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 3]
```diff
@@ -2873,6 +2873,354 @@ static void Main()
...
+                    C.M(in x);
+                }
+                static void M2() => C.M();
+            }
+            """;
+        var comp = fromMetadata
+            ? CreateCompilation(source2, new[] { CreateCompilation(source1).ToMetadataReference() }, options: TestOptions.ReleaseExe)
+            : CreateCompilation(new[] { source1, source2 }, options: TestOptions.ReleaseExe);
+        var verifier = CompileAndVerify(comp, expectedOutput: "1222").VerifyDiagnostics(
+            // (3,47): warning CS9521: A default value is specified for 'ref readonly' parameter 'i', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2903]: >// (3,47): warning CS9521: A default value is specified for 'ref readonly' parameter 'i', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'. [](http://example.com/codeflow?start=12&length=186)

Is this warning expected when ```fromMetadata``` is true? And if its, then why?

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2988]: https://github.com/dotnet/roslyn/issues/69078

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 4]
```diff
@@ -2873,6 +2873,365 @@ static void Main()
...
+                }
+                static void M2() => C.M();
+            }
+            """;
+        var warning2 =
+            // (7,13): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
+            //         C.M(x);
+            Diagnostic(ErrorCode.WRN_ArgExpectedRefOrIn, "x").WithArguments("1").WithLocation(7, 13);
+        var comp = fromMetadata
+            ? CreateCompilation(source2, new[] { CreateCompilation(source1).VerifyDiagnostics(warning1).EmitToImageReference() }, options: TestOptions.ReleaseExe).VerifyDiagnostics(warning2)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2908]: >VerifyDiagnostics(warning1) [](http://example.com/codeflow?start=76&length=27)

I think verifying this diagnostics isn't that interesting.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 5]
```diff
@@ -2873,6 +2873,365 @@ static void Main()
...
+            : CreateCompilation(new[] { source1, source2 }, options: TestOptions.ReleaseExe).VerifyDiagnostics(warning1, warning2);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:2909]: >VerifyDiagnostics(warning1, warning2) [](http://example.com/codeflow?start=93&length=37)

I think that we should verify diagnostics on ```CompileAndVerify``` instead. The set of diagnostics will be conditional on ```fromMetadata```, but we want to verify complete set of diagnostics, including from emit phase for the compilation.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:5]: >System [](http://example.com/codeflow?start=6&length=6)

Is this using necessary?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4), assuming CI is passing

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs (hunk 6 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs:5]: Yes, for Array.Empty for conditional diagnostics in one test

**Pull request merged**